### PR TITLE
FAP: Introduce some metrics and enhancement

### PR DESCRIPTION
**Pull request by CalvinNeo**
### What problem does this PR solve?

Issue Number: ref #8265

Problem Summary:

image

FAP time by range -> total previously got some problems, but fixed then
image

image

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Storages/DeltaMerge/Segment.cpp [hunk 1]
```diff
@@ -471,8 +472,10 @@ Segments Segment::createTargetSegmentsFromCheckpoint( //
             segment_info.epoch,
             segment_info.next_segment_id);
         auto stable = StableValueSpace::createFromCheckpoint(context, temp_ps, segment_info.stable_id, wbs);
+        LOG_INFO(parent_log, "!!!!! afterr stable {}", wbs.data.empty());
         auto delta
             = DeltaValueSpace::createFromCheckpoint(context, temp_ps, segment_info.range, segment_info.delta_id, wbs);
+        LOG_INFO(parent_log, "!!!!! afterr delta {}", wbs.data.empty());
```
[the remark this answers is not in the document]
Comment (CalvinNeo) [reply dbms/src/Storages/DeltaMerge/Segment.cpp:478]: Sorry, these shoulbe be removed...

File: dbms/src/Common/TiFlashMetrics.h [hunk 2]
```diff
@@ -372,6 +372,32 @@ namespace DB
...
+      F(type_failed_other, {{"type", "failed_other"}}),                                                                             \
+      F(type_failed_cancel, {{"type", "failed_cancel"}}),                                                                           \
+      F(type_failed_timeout, {{"type", "failed_timeout"}}),                                                                         \
+      F(type_succeed, {{"type", "succeed"}}))                                                                                       \
+    M(tiflash_fap_task_state,                                                                                                       \
+      "",                                                                                                                           \
+      Gauge,                                                                                                                        \
+      F(type_ongoing, {{"type", "ongoing"}}),                                                                                       \
```
[the remark this answers is not in the document]
Comment (CalvinNeo) [reply dbms/src/Common/TiFlashMetrics.h:386]: ongoing = building + ingesting + queueing

File: dbms/src/Common/TiFlashMetrics.h [hunk 3]
```diff
@@ -372,6 +372,32 @@ namespace DB
...
+      F(type_conf, {{"type", "conf"}}),                                                                                             \
+      F(type_region_state, {{"type", "region_state"}}),                                                                             \
+      F(type_no_meta, {{"type", "no_meta"}}))                                                                                       \
+    M(tiflash_fap_task_duration_seconds,                                                                                            \
+      "",                                                                                                                           \
+      Histogram,                                                                                                                    \
+      F(type_build, {{"type", "build"}}, ExpBuckets{0.05, 2, 60}),                                                                  \
+      F(type_success, {{"type", "success"}}, ExpBuckets{0.05, 2, 60}),                                                              \
```
[the remark this answers is not in the document]
Comment (CalvinNeo) [reply dbms/src/Common/TiFlashMetrics.h:398]: total = success + failed
success = ingest + build + queueing
failed = sum(different failed reasons)

File: dbms/src/Common/TiFlashMetrics.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (CalvinNeo) [reply dbms/src/Common/TiFlashMetrics.h:386]: I will add a `_stage` suffix to indicate this is a sub process of fap

File: dbms/src/Common/TiFlashMetrics.h [hunk 4]
```diff
@@ -372,6 +372,33 @@ namespace DB
...
+      F(type_failed_cancel, {{"type", "failed_cancel"}}),                                                                           \
+      F(type_failed_timeout, {{"type", "failed_timeout"}}),                                                                         \
+      F(type_succeed, {{"type", "succeed"}}))                                                                                       \
+    M(tiflash_fap_task_state,                                                                                                       \
+      "",                                                                                                                           \
+      Gauge,                                                                                                                        \
+      F(type_ongoing, {{"type", "ongoing"}}),                                                                                       \
+      F(type_ingesting_stage, {{"type", "ingesting_stage"}}),                                                                       \
```
Comment (CalvinNeo) [inline dbms/src/Common/TiFlashMetrics.h:387]: It is meaningful if we switch to two phase fap later.

File: dbms/src/Storages/KVStore/tests/gtest_new_kvstore.cpp [hunk 5]
```diff
@@ -686,6 +687,65 @@ try
...
+        auto count = std::accumulate(f.begin(), f.end(), 0, [&](int a, bool b) -> int { return a + int(b); });
+        if (count >= total)
+        {
+            break;
+        }
+        else
+        {
+            LOG_DEBUG(log, "finished {}/{}", count, total);
+        }
+        for (int i = 0; i < total; i++)
```
Comment (Lloyd-Pottiger) [inline dbms/src/Storages/KVStore/tests/gtest_new_kvstore.cpp:712]:
```suggestion
        for (int i = 0; i < total; ++i)
```

File: dbms/src/Storages/KVStore/tests/gtest_new_kvstore.cpp [hunk 6]
```diff
@@ -686,6 +687,65 @@ try
...
+                if (initial_loop)
+                {
+                    if (i <= 1)
+                    {
+                        ASSERT_EQ(res, true);
+                    }
+                    else
+                    {
+                        ASSERT_EQ(res, false);
+                    }
```
Comment (Lloyd-Pottiger) [inline dbms/src/Storages/KVStore/tests/gtest_new_kvstore.cpp:729]:
```suggestion
                    ASSERT_EQ(res, I <= 1);
```

File: dbms/src/Storages/KVStore/tests/gtest_new_kvstore.cpp [hunk 7]
```diff
@@ -686,6 +687,65 @@ try
...
+                }
+            }
+        }
+
+        for (int i = 0; i < total; i++)
```
Comment (Lloyd-Pottiger) [inline dbms/src/Storages/KVStore/tests/gtest_new_kvstore.cpp:734]:
```suggestion
        for (int i = 0; i < total; ++i)
```

**Review by Lloyd-Pottiger — APPROVED** [review]
Others LGTM

**Pull request merged**